sql: panic: interface conversion: coldata.column is []int16, not []int64

```
SET experimental_vectorize = "on";
CREATE TABLE t (_int2) AS SELECT 1::INT2;
SELECT sum_int(t._int2) FROM t;
```

```
panic: interface conversion: coldata.column is []int16, not []int64 [recovered]
	panic: panic while executing 1 statements: SELECT sum_int(_._) FROM _; caused by interface conversion: coldata.column is []int16, not []int64

goroutine 441 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000908d80, 0x43a66c0, 0xc000ff88c0, 0x3738fc0, 0xc001297410)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000908d80, 0x43a66c0, 0xc000ff88c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x3738fc0, 0xc001297410)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Int64(0xc0013feac0, 0x0, 0x4407e00, 0xc0013feac0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:210 +0x74
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc0007aa700, 0x400, 0x400, 0xc000e64230, 0x2514f7e)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:161 +0x3ce
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x43a6780, 0xc001296a80, 0x43b2440, 0xc0007aa700, 0x4383540, 0xc00180e2c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc0007aa700, 0x43a6780, 0xc001296a80)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00000cf00, 0x43a6780, 0xc001296a80, 0x3c9b140, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0001ca5a0, 0xc0015b6300, 0xc0012350e0, 0xc000e64870, 0xc00180e2c0, 0xc0009091a8, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0001ca5a0, 0x43a6780, 0xc000a969f0, 0xc0009091a8, 0xc0015b6300, 0xc0012350e0, 0x43a8000, 0xc000721f00, 0xc00180e2c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000908d80, 0x43a6780, 0xc000a969f0, 0xc0009090d0, 0x3, 0x7fd77f375f68, 0xc000683760, 0xc000e64a01, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:902 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000908d80, 0x43a6780, 0xc000a969f0, 0xc0009090d0, 0x7fd77f375f68, 0xc000683760, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:734 +0x6e5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000908d80, 0x43a6780, 0xc000a969f0, 0x43abc80, 0xc0006f0740, 0xc00177a20b, 0x1e, 0x0, 0x1, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000908d80, 0x43a6780, 0xc000a969f0, 0x43abc80, 0xc0006f0740, 0xc00177a20b, 0x1e, 0x0, 0x1, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000908d80, 0x43a66c0, 0xc0006f0780, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1204 +0x1ab0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000908d80, 0x43a66c0, 0xc000ff88c0, 0xc00087c7c0, 0x5400, 0x15000, 0xc00087c858, 0xc00033a9a0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0001c1680, 0x43a66c0, 0xc000ff88c0, 0xc000908d80, 0x5400, 0x15000, 0xc00087c858, 0xc00033a9a0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc00038cf79, 0xc0016a40c0, 0x43a66c0, 0xc000ff88c0, 0xc00033a9a0, 0xc0001c1680, 0xc00179a000, 0x43aa200, 0xc0016a40a0, 0xc00090ea80, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:509 +0x17b
```

--- a/pkg/sql/distsqlrun/column_exec_setup.go
+++ b/pkg/sql/distsqlrun/column_exec_setup.go
@@ -204,6 +204,11 @@ func newColOperator(
 					// issues, at first, we could plan SUM for all types besides Int64.
 					return nil, nil, memUsage, errors.Newf("sum on int cols not supported (use sum_int)")
 				}
+			case distsqlpb.AggregatorSpec_SUM_INT:
+				// TODO(yuzefovich): support this case through vectorize.
+				if aggTyps[i][0].Width() != 64 {
+					return nil, nil, memUsage, errors.Newf("sum_int is only supported on Int64 through vectorized")
+				}
 			}
 			_, retType, err := GetAggregateInfo(agg.Func, aggTyps[i]...)
 			if err != nil {

--- a/pkg/sql/distsqlrun/materializer.go
+++ b/pkg/sql/distsqlrun/materializer.go
@@ -46,6 +46,11 @@ type materializer struct {
 
 	// row is the memory used for the output row.
 	row sqlbase.EncDatumRow
+
+	// Fields to store the returned results of next() to be passed through an
+	// adapter.
+	outputRow      sqlbase.EncDatumRow
+	outputMetadata *distsqlpb.ProducerMetadata
 }
 
 func newMaterializer(
@@ -95,21 +100,20 @@ func (m *materializer) Start(ctx context.Context) context.Context {
 	return ctx
 }
 
-// nextBatch saves the next batch from input in m.batch. For internal use only.
-// The purpose of having this function is to not create an anonymous function
-// on every call to Next().
-func (m *materializer) nextBatch() {
-	m.batch = m.input.Next(m.Ctx)
+// nextAdapter calls next() and saves the returned results in m. For internal
+// use only. The purpose of having this function is to not create an anonymous
+// function on every call to Next().
+func (m *materializer) nextAdapter() {
+	m.outputRow, m.outputMetadata = m.next()
 }
 
-func (m *materializer) Next() (sqlbase.EncDatumRow, *distsqlpb.ProducerMetadata) {
+// next is the logic of Next() extracted in a separate method to be used by an
+// adapter to be able to wrap the latter with a catcher.
+func (m *materializer) next() (sqlbase.EncDatumRow, *distsqlpb.ProducerMetadata) {
 	for m.State == StateRunning {
 		if m.batch == nil || m.curIdx >= m.batch.Length() {
 			// Get a fresh batch.
-			if err := exec.CatchVectorizedRuntimeError(m.nextBatch); err != nil {
-				m.MoveToDraining(err)
-				return nil, m.DrainHelper()
-			}
+			m.batch = m.input.Next(m.Ctx)
 
 			if m.batch.Length() == 0 {
 				m.MoveToDraining(nil /* err */)
@@ -143,6 +147,14 @@ func (m *materializer) Next() (sqlbase.EncDatumRow, *distsqlpb.ProducerMetadata)
 		return m.ProcessRowHelper(m.row), nil
 	}
 	return nil, m.DrainHelper()
+}
+
+func (m *materializer) Next() (sqlbase.EncDatumRow, *distsqlpb.ProducerMetadata) {
+	if err := exec.CatchVectorizedRuntimeError(m.nextAdapter); err != nil {
+		m.MoveToDraining(err)
+		return nil, m.DrainHelper()
+	}
+	return m.outputRow, m.outputMetadata
 }
 
 func (m *materializer) ConsumerClosed() {

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -515,3 +515,12 @@ query I
 SELECT EXTRACT(YEAR FROM x) FROM extract_test
 ----
 2017
+
+# Regression test for #38937
+statement ok
+CREATE TABLE t38937 (_int2) AS SELECT 1::INT2
+
+query I
+SELECT sum_int(_int2) FROM t38937
+----
+1

